exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,398 @@
+    /// #
+    /// use exonum::blockchain::Transaction;
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::messages::Message;
+    /// # use exonum::blockchain::ExecutionStatus;
+    /// # use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #
+    ///         public_key: &PublicKey,
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn verify(&self) -> bool {
+    ///         self.verify_signature(self.public_key())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn execute(&self, _: &mut Fork) -> ExecutionStatus {
+    /// #       Ok(())
+    /// #   }
+    /// }
+    /// # fn main() {}
+    fn verify(&self) -> bool;
+
+    /// Receives a fork of the current blockchain state and can modify it depending on the contents
+    /// of the transaction.
+    ///
+    /// # Notes
+    ///
+    /// - Transaction itself is considered committed regardless whether `Ok` or `Err` has been
+    ///   returned or even if panic occurred during execution.
+    /// - Changes made by the transaction are discarded if `Err` is returned or panic occurred.
+    /// - A transaction execution status (see `ExecutionStatus` and `TransactionStatus` for the
+    ///   details) is stored in the blockchain and can be accessed through api.
+    /// - Blockchain state hash is affected by the transactions execution status.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::{Transaction, ExecutionStatus};
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #
+    /// #       public_key: &PublicKey,
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> ExecutionStatus {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Return execution status.
+    ///         Ok(())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> ExecutionStatus;
+}
+
+/// Result of unsuccessful transaction execution.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExecutionError {
+    /// General failure (unspecified reason).
+    Failure,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+    /// User-defined string error description.
+    Description(String),
```

Comment:
Having both `Code` and `Description` variants seems simultaneously overly restrictive and overly general. Like, you can't do much with code, but you can store arbitrary JSON in description if you want too...

Reply:
I suggest `Description(Box<Error>)`

Reply:
@matklad 
> Like, you can't do much with code

Your can store (almost) anything in the `MapIndex`. :wink: 

The original idea was the following: `Code` gives you enough flexibility to organize the error handling as complex as you want. Additionally, it is easier to deal with codes having versioning in mind. String descriptions are quite the opposite: they are nice when you only need to report error to user, but are harder to maintain. I think that both variants can be useful.